source/common/router/retry_state_impl.cc

```diff
@@ -147,6 +166,8 @@ uint32_t RetryStateImpl::parseRetryGrpcOn(absl::string_view retry_grpc_on_header
       ret |= RetryPolicy::RETRY_ON_GRPC_UNAVAILABLE;
     } else if (retry_on == Http::Headers::get().EnvoyRetryOnGrpcValues.Internal) {
       ret |= RetryPolicy::RETRY_ON_GRPC_INTERNAL;
+    } else if (flag_parse_failures) {
+      ret |= RetryPolicy::RETRY_UNKNOWN_FIELD_ERROR;
```

Comment:
I had initially tried this approach, but it turns out to be unsafe because we reuse `RetryStateImpl::{parseRetryOn, parseGrpcRetryOn}` to parse the `retry_on` config field for `route.RetryPolicy` ([docs](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto#route-retrypolicy)):
> retry_on
(string) Specifies the conditions under which retry takes place. These are the same conditions documented for x-envoy-retry-on and x-envoy-retry-grpc-on.

The code _unions_ the flags from each parsing function here: https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.cc#L73-L74

So this change makes a `strict` version of each parsing function.

Reply:
> Perhaps I can make the strict versions call the non-strict versions, and if the non-strict parsers return 0, always set RETRY_UNKNOWN_FIELD_ERROR. WDYT?

Hrmm. Actually this won't work for invalid values e.g. `"cancelled,deadline-exceeded,NOT_VALID"`

Reply:
More context from slack + decision with snowp:
- I initially separated this logic because this [test](https://github.com/envoyproxy/envoy/blob/cc0d89a8624dfff4fe00949fab697cc0d15a9cb4/test/common/router/config_impl_test.cc#L2336-L2340
) sets a `retry_on` config to `5xx,gateway-error,connect-failure` (a mixed set of retry policies), which would fail if we unconditionally set `RETRY_UNKNOWN_FIELD_ERROR`

We'll simplify this by unconditionally clearing this flag where this [config option is parsed](https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.cc#L73-L74).